src/foraging_gui/Foraging.py

```diff
@@ -525,12 +517,58 @@ def _GetSettings(self):
             self.Tower.setCurrentIndex(index)
 
     def _InitializeBonsai(self):
-        '''Initializing osc messages'''
-        logging.info('initializing Bonsai')
-        # open the bondai workflow and run
+        '''
+            Connect to Bonsai using OSC messages to establish a connection. 
+            
+            We first attempt to connect, to see if Bonsai is already running. 
+            If not, we start Bonsai and check the connection every 500ms.
+            If we wait more than 30 seconds without Bonsai connection we set 
+            InitializeBonsaiSuccessfully=0 and return
+    
+        '''
+
+        # Try to connect, to see if Bonsai is already runniny
+        try: 
+            logging.info('Trying to connect to already running Bonsai')
+            self._ConnectOSC()
+        except Exception as e:
+            # We couldn't connect, log as info, and move on
+            logging.info('Could not connect: '+str(e))
+        else:
+            # We could connect, set the indicator flag and return
+            logging.info('Connected to already running Bonsai')
+            self.InitializeBonsaiSuccessfully=1
+            logging.info('Bonsai started successfully')
+            return
+
+        # Start Bonsai
+        logging.info('Starting Bonsai')
         self._OpenBonsaiWorkflow()
-        time.sleep(3)
-        self._ConnectOSC()
+
+        # Test the connection until it completes or we time out
+        wait = 0
+        max_wait = 30
+        check_every = .5
+        while wait < max_wait:
+            time.sleep(check_every)
+            wait += check_every
+            try:
+                self._ConnectOSC()
+            except Exception as e:
+                # We could not connect
+                logging.info('Could not connect, total waiting time {} seconds: '.format(wait)+str(e))
+            else:
+                # We could connect
+                logging.info('Connected to Bonsai after {} seconds'.format(wait))               
+                self.InitializeBonsaiSuccessfully=1
+                logging.info('Bonsai started successfully')
+                return
```

Comment:
The GUI will freeze during these 30 seconds, which is not what we want. We can set max_wait = 3. Normally, it will be completed within 3s, otherwise we can establish the connection manually.

Reply:
Yes, the GUI will freeze during those 30 seconds, but only if Bonsai has not opened. In the last two weeks 25% of all GUI sessions did not connect in the 3 seconds. So we could set the max_wait to be smaller than 30, but I think it should be greater than 3s. 

What do you mean by establish the connection manually? Do you mean establish the connection later?